Derivs/src/derivs_spacetimex.hxx

```diff
@@ -0,0 +1,512 @@
+deriv(const simdl<T> &mask, const GF3D2<const T> &gf_, const vect<int, dim> &I,
+      const vec<T, dim> &dx) {
+  return {deriv<0>(mask, gf_, I, dx), deriv<1>(mask, gf_, I, dx),
+          deriv<2>(mask, gf_, I, dx)};
+}
+
+template <typename T>
+inline ARITH_INLINE ARITH_DEVICE ARITH_HOST simd<T>
+deriv2_1d(const simdl<T> &mask, const T *restrict const var, const ptrdiff_t di,
+          const T dx) {
+  const auto load = [&](const int n) {
+    return maskz_loadu(mask, &var[n * di]) + maskz_loadu(mask, &var[-n * di]);
+  };
+  const auto load0 = [&]() { return maskz_loadu(mask, &var[0]); };
+  if constexpr (deriv_order == 2)
+    return (load(1) - 2 * load0()) / pow2(dx);
+  if constexpr (deriv_order == 4)
+    return (1 / T(12) * load(2) - 4 / T(3) * load(1) + 5 / T(2) * load0()) /
+           pow2(dx);
+  if constexpr (deriv_order == 6)
+    return (1 / T(90) * load(3) - 3 / T(20) * load(2) + 3 / T(2) * load(1) -
+            49 / T(18) * load0()) /
+           pow2(dx);
+}
+
+using std::tuple_size_v;
+using Arith::min;
+using std::array;
+using Arith::div_floor;
+using Arith::div_ceil;
+using Arith::align_ceil;
+using Arith::mask_for_loop_tail;
+
+template <typename T>
+inline ARITH_INLINE ARITH_DEVICE ARITH_HOST simd<T>
+deriv2_2d(const int vavail, const simdl<T> &mask, const T *restrict const var,
+          const ptrdiff_t di, const ptrdiff_t dj, const T dx, const T dy) {
+  constexpr size_t vsize = tuple_size_v<simd<T> >;
+  if (di == 1) {
+    assert(vavail > 0);
+    constexpr int maxnpoints = deriv_order + 1 + vsize - 1;
+    const int npoints = deriv_order + 1 + min(int(vsize), vavail) - 1;
+    array<simd<T>, div_ceil(maxnpoints, int(vsize))> arrx;
+    for (int i = 0; i < maxnpoints; i += vsize) {
+      if (i < npoints) {
+        const simdl<T> mask1 = Arith::mask_for_loop_tail<simdl<T> >(i, npoints);
+        arrx[div_floor(i, int(vsize))] =
+            deriv1d(mask1, &var[i - deriv_order / 2], dj, dy);
+      }
+    }
+#ifdef CCTK_DEBUG
+    for (int i = npoints; i < align_ceil(maxnpoints, int(vsize)); ++i)
+      ((T *)&arrx[0])[i] = Arith::nan<T>()(); // unused
+#endif
+    const T *const varx = (T *)&arrx[0] + deriv_order / 2;
+    return deriv1d(mask, varx, 1, dx);
+  } else {
+    assert(dj != 1);
+    array<simd<T>, deriv_order + 1> arrx;
+    for (int j = -deriv_order / 2; j <= deriv_order / 2; ++j)
+      if (j == 0) {
+#ifdef CCTK_DEBUG
+        arrx[deriv_order / 2 + j] = Arith::nan<simd<T> >()(); // unused
+#endif
+      } else {
+        arrx[deriv_order / 2 + j] = deriv1d(mask, &var[j * dj], di, dx);
+      }
+    const T *const varx = (T *)(&arrx[deriv_order / 2]);
+    return deriv1d(mask, varx, vsize, dy);
+  }
+}
+
+using Loop::GF3D5;
+using Loop::GF3D5layout;
+using Loop::GF3D5index;
+using Arith::smat;
+using Loop::PointDesc;
+
+
+template <typename T>
+CCTK_ATTRIBUTE_NOINLINE void
+calc_derivs(const cGH *restrict const cctkGH, const GF3D2<const T> &gf1,
```

Comment:
The input variables are called `...1`, and the output variables `...0`. That's a weird scheme. Maybe call the input quantities `...0`, and use no suffix for the output quantities?